### DISCUSSION TITLE
using `.patch` files

### DISCUSSION THREAD

**Discussion by mainrs** [libgit2/libgit2#5588] [General] [answered]
Hey! 

Can I use libgit2 with `.patch` files that can be applied to repositories? If so, how exactly? I have a tool which is configurable and depending on what the user wants, I want to sequentially apply different patches for some files.

Sadly I can't rely on system commands like `patch`.

**Comment by ethomson** [comment, marked as answer]
You can `git_apply` a `diff` to a repository using [`git_apply`](https://github.com/libgit2/libgit2/blob/master/include/git2/apply.h#L141).  You can get a `git_diff` from a patch file with [`git_diff_from_buffer `](https://github.com/libgit2/libgit2/blob/master/include/git2/diff.h#L1253).

*However*, our diff machinery is built to read patch files that were produced by Git implementations (git, libgit2, jgit, etc).  We don't necessarily support a generic patch application - patch files are extraordinarily wide-ranging in their format.  So if you have something that looks like a git diff, great!  If not, we may not be able to parse it.

**Reply by mainrs** [reply]
Thanks for the help! I did skim through the documentation but didn't see `git_diff_from_buffer` 👍
